src/sbvr-api/sbvr-utils.coffee

```diff
@@ -687,34 +765,38 @@ exports.handleODataRequest = handleODataRequest = (req, res, next) ->
 		# Parse the OData requests
 		mapSeries body, (bodypart) ->
 			uriParser.parseOData(bodypart)
-			.then controlFlow.liftP (request) ->
-				# Get the full hooks list now that we can. We clear the hooks on
-				# the global req to avoid duplication
-				req.hooks = {}
-				request.hooks = getHooks(request)
-				# Add/check the relevant permissions
-				runHook('POSTPARSE', { req, request, tx: req.tx })
-				.return(request)
-				.then (uriParser.translateUri)
-				.then (request) ->
-					# We defer compilation of abstract sql queries with references to other requests
-					if request.abstractSqlQuery? && !request._defer
-						try
-							request.sqlQuery = memoizedCompileRule(request.abstractSqlQuery)
-						catch err
-							api[apiRoot].logger.error('Failed to compile abstract sql: ', request.abstractSqlQuery, err, err.stack)
-							throw new SqlCompilationError(err)
-					return request
-
-			.then (request) ->
-				# Run the request in its own transaction
-				runTransaction req, (tx) ->
-					if _.isArray request
-						env = new Map()
-						Promise.reduce(request, runChangeSet(req, res, tx), env)
-						.then (env) -> Array.from(env.values())
-					else
-						runRequest(req, res, tx, request)
+			.then (odata) ->
+				requestEnv = prepareEnv(req, odata)
+				Promise.resolve(requestEnv)
+				.then controlFlow.liftP (requestEnv) ->
+					{ req, request } = requestEnv
+					# Get the full hooks list now that we can, we overwrite the previous list to avoid duplication
+					request.hooks = getHooks(request)
```

Comment:
There is an issue atm when refreshing the hook list, since we now rely on the executed flag to be set on the hook to rollback. 
This is not a problem as there is no PREPARSE hook that does any side effect action, nor the probably should ever be. Maybe we can prevent a PREPARSE side-effect hook from ever being created by checking in `addSideEffectHook` just to be sure.